remove caching at github client layer in favor of httpcli caching

davejrt opened the pull request:
Removes the use of a cache at the github client layer for https://github.com/sourcegraph/sourcegraph/issues/30656. We take advantage of caching in the httpcli package, specifically [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/httpcli/client.go?L319&subtree=true) which uses [this](https://sourcegraph.com/github.com/gregjones/httpcache/-/blob/httpcache.go?L139) function from the httpcache package. 

GetRepository has been updated to use vcr tests. 

Follow up will be to migrate other tests as part of https://github.com/sourcegraph/sourcegraph/issues/33839

## Test plan

Tested locally, and new keys can be seen here in the redis cache (keys 7, 19 and 23), with the second test call taking less time due to a cache hit. 
```

127.0.0.1:6379> KEYS *
➜  github git:(dt/use_etag) ✗ go test -v -run TestListMembers 
=== RUN   TestListMembers
=== RUN   TestListMembers/org_members
.....DAVE GET......
DBUG[03-30|09:57:00] TRACE github                             host=api.github.com path=/orgs/sourcegraph-vcr-repos/members code=200 duration=498.438042ms
<<<<<<DAVE SET>>>>>>
.....DAVE GET......
DBUG[03-30|09:57:00] TRACE github                             host=api.github.com path=/orgs/sourcegraph-vcr-repos/members code=200 duration=6.366042ms
=== RUN   TestListMembers/org_admins
.....DAVE GET......
DBUG[03-30|09:57:01] TRACE github                             host=api.github.com path=/orgs/sourcegraph-vcr-repos/members code=200 duration=175.071084ms
<<<<<<DAVE SET>>>>>>
=== RUN   TestListMembers/team_members
.....DAVE GET......
DBUG[03-30|09:57:01] TRACE github                             host=api.github.com path=/orgs/sourcegraph-vcr-repos/teams/private-access/members code=200 duration=222.961959ms
<<<<<<DAVE SET>>>>>>
--- PASS: TestListMembers (0.96s)
    --- PASS: TestListMembers/org_members (0.53s)
    --- PASS: TestListMembers/org_admins (0.19s)
    --- PASS: TestListMembers/team_members (0.24s)
PASS
ok      github.com/sourcegraph/sourcegraph/internal/extsvc/github       1.324s
➜  github git:(dt/use_etag) ✗ docker exec -ti dev-redis-1 redis-cli 
127.0.0.1:6379> KEYS * 
1) "v2:http:https://api.github.com/orgs/sourcegraph-vcr-repos/teams/private-access/members?page=1&per_page=100"
2) "v2:http:https://api.github.com/orgs/sourcegraph-vcr-repos/members?page=1&per_page=100"
3) "v2:http:https://api.github.com/orgs/sourcegraph-vcr-repos/members?page=1&per_page=100&role=admin"
127.0.0.1:6379> 
```

Update to GetRepository to directly use the cache as well with a test below

```
➜  github git:(dt/use_etag) ✗ docker exec -ti dev-redis-1 redis-cli 
127.0.0.1:6379> KEYS * 
(empty list or set)
127.0.0.1:6379> exit
➜  github git:(dt/use_etag) ✗ go test -run TestClient_GetRepository
DBUG[04-12|17:32:44] TRACE github                             host=api.github.com path=/repos/sourcegraph-vcr-repos/private-org-repo-1 code=200 duration=524.473958ms
DBUG[04-12|17:32:44] TRACE github                             host=example.com path=/repos/owner/repo code=404 duration=1.75µs
PASS
ok      github.com/sourcegraph/sourcegraph/internal/extsvc/github       0.913s
➜  github git:(dt/use_etag) ✗ docker exec -ti dev-redis-1 redis-cli 
127.0.0.1:6379> KEYS * 
1) "v2:http:https://api.github.com/repos/sourcegraph-vcr-repos/private-org-repo-1"
127.0.0.1:6379> 
```

davejrt commented:
Coming up against some test failures here that I just can't seem to figure out. 

Some seem to be token related: for example locally
```
➜  repos git:(dt/use_etag) ✗ go test -v -run TestRepositoryQuery_Do
=== RUN   TestRepositoryQuery_Do
=== RUN   TestRepositoryQuery_Do/exceeds-limit
DBUG[03-31|19:45:43] TRACE github                             host=api.github.com path=/graphql code=200 duration=771.827083ms
INFO[03-31|19:45:43] repositoryQuery matched more than 20 results, refining it and retrying query=stars:10000..10100
INFO[03-31|19:45:43] repositoryQuery refined                  query="stars:10000..10100 created:2007-06-01T00:00:00+00:00..2022-03-31T19:45:43+00:00"
DBUG[03-31|19:45:44] TRACE github                             host=api.github.com path=/graphql code=200 duration=684.790042ms
INFO[03-31|19:45:44] repositoryQuery matched more than 20 results, refining it and retrying query="stars:10000..10100 created:2007-06-01T00:00:00+00:00..2022-03-31T19:45:43+00:00"
INFO[03-31|19:45:44] repositoryQuery refined                  query="stars:10000..10100 created:2014-10-30T21:52:51+00:00..2022-03-31T19:45:43+00:00"
DBUG[03-31|19:45:44] TRACE github                             host=api.github.com path=/graphql code=200 duration=483.635584ms
INFO[03-31|19:45:44] repositoryQuery matched more than 20 results, refining it and retrying query="stars:10000..10100 created:2014-10-30T21:52:51+00:00..2022-03-31T19:45:43+00:00"
INFO[03-31|19:45:44] repositoryQuery refined                  query="stars:10000..10100 created:2018-07-16T08:49:17+00:00..2022-03-31T19:45:43+00:00"
DBUG[03-31|19:45:44] TRACE github                             host=api.github.com path=/graphql code=200 duration=305.59575ms
INFO[03-31|19:45:44] repositoryQuery matched less than 10 results, expanding it and retrying query="stars:10000..10100 created:2018-07-16T08:49:17+00:00..2022-03-31T19:45:43+00:00"
INFO[03-31|19:45:44] repositoryQuery expanded                 query="stars:10000..10100 created:2016-09-07T03:21:04+00:00..2022-03-31T19:45:43+00:00"
DBUG[03-31|19:45:45] TRACE github                             host=api.github.com path=/graphql code=200 duration=420.232542ms
INFO[03-31|19:45:45] repositoryQuery matched less than 10 results, expanding it and retrying query="stars:10000..10100 created:2016-09-07T03:21:04+00:00..2022-03-31T19:45:43+00:00"
INFO[03-31|19:45:45] repositoryQuery expanded                 query="stars:10000..10100 created:2013-11-26T07:08:45+00:00..2022-03-31T19:45:43+00:00"
DBUG[03-31|19:45:45] TRACE github                             host=api.github.com path=/graphql code=200 duration=406.443167ms
INFO[03-31|19:45:45] repositoryQuery matched more than 20 results, refining it and retrying query="stars:10000..10100 created:2013-11-26T07:08:45+00:00..2022-03-31T19:45:43+00:00"
INFO[03-31|19:45:45] repositoryQuery refined                  query="stars:10000..10100 created:2018-01-28T01:27:14+00:00..2022-03-31T19:45:43+00:00"
DBUG[03-31|19:45:46] TRACE github                             host=api.github.com path=/graphql code=200 duration=341.85425ms
INFO[03-31|19:45:46] repositoryQuery matched less than 10 results, expanding it and retrying query="stars:10000..10100 created:2018-01-28T01:27:14+00:00..2022-03-31T19:45:43+00:00"
INFO[03-31|19:45:46] repositoryQuery expanded                 query="stars:10000..10100 created:2015-12-28T04:17:59+00:00..2022-03-31T19:45:43+00:00"
DBUG[03-31|19:45:46] TRACE github                             host=api.github.com path=/graphql code=200 duration=385.071916ms
INFO[03-31|19:45:46] repositoryQuery matched                  query="stars:10000..10100 created:2015-12-28T04:17:59+00:00..2022-03-31T19:45:43+00:00" total=13 page=10
DBUG[03-31|19:45:46] TRACE github                             host=api.github.com path=/graphql code=200 duration=254.715084ms
INFO[03-31|19:45:46] repositoryQuery matched                  query="stars:10000..10100 created:2015-12-28T04:17:59+00:00..2022-03-31T19:45:43+00:00" total=13 page=3
DBUG[03-31|19:45:47] TRACE github                             host=api.github.com path=/graphql code=200 duration=427.85875ms
INFO[03-31|19:45:47] repositoryQuery matched                  query="stars:10000..10100 created:2009-09-24T12:50:15+00:00..2015-12-28T04:17:58+00:00" total=15 page=10
DBUG[03-31|19:45:47] TRACE github                             host=api.github.com path=/graphql code=200 duration=316.490959ms
INFO[03-31|19:45:47] repositoryQuery matched                  query="stars:10000..10100 created:2009-09-24T12:50:15+00:00..2015-12-28T04:17:58+00:00" total=15 page=5
DBUG[03-31|19:45:47] TRACE github                             host=api.github.com path=/graphql code=200 duration=189.25375ms
INFO[03-31|19:45:47] repositoryQuery matched less than 10 results, expanding it and retrying query="stars:10000..10100 created:2007-06-01T00:00:00+00:00..2009-09-24T12:50:14+00:00"
INFO[03-31|19:45:47] repositoryQuery matched                  query="stars:10000..10100 created:2007-06-01T00:00:00+00:00..2009-09-24T12:50:14+00:00" total=0 page=0
=== RUN   TestRepositoryQuery_Do/doesnt-exceed-limit
DBUG[03-31|19:45:47] TRACE github                             host=api.github.com path=/graphql code=200 duration=200.053042ms
INFO[03-31|19:45:47] repositoryQuery matched less than 10 results, expanding it and retrying query="repo:tsenart/vegeta stars:>=14000"
INFO[03-31|19:45:47] repositoryQuery matched                  query="repo:tsenart/vegeta stars:>=14000" total=1 page=1
--- PASS: TestRepositoryQuery_Do (5.21s)
    --- PASS: TestRepositoryQuery_Do/exceeds-limit (5.01s)
    --- PASS: TestRepositoryQuery_Do/doesnt-exceed-limit (0.20s)
PASS
ok      github.com/sourcegraph/sourcegraph/internal/repos       5.643s
```

```
➜  resolvers git:(dt/use_etag) go test -run TestChangesetCountsOverTimeIntegration
2022/03/31 20:07:06 logging complete query for unnamed GraphQL request above name=unknown userID=1 source=other:
QUERY
-----

query($batchChange: ID!, $from: DateTime!, $to: DateTime!) {
  node(id: $batchChange) {
    ... on BatchChange {
    changesetCountsOverTime(from: $from, to: $to) {
        date
        total
        merged
        draft
        closed
        open
        openApproved
        openChangesRequested
        openPending
      }
    }
  }
}


VARIABLES
---------
map[batchChange:QmF0Y2hDaGFuZ2U6MQ== from:2019-10-01T14:49:31Z to:2019-10-07T13:13:45Z]

PASS
ok  	github.com/sourcegraph/sourcegraph/enterprise/cmd/frontend/internal/batches/resolvers	6.591s
```

bobheadxi commented on internal/httptestutil/recorder.go:
```diff
@@ -77,7 +79,7 @@ func NewGitHubRecorderFactory(t testing.TB, update bool, name string) (*httpcli.
 	mw := httpcli.NewMiddleware(httpcli.GitHubProxyRedirectMiddleware)
 
-	hc := httpcli.NewFactory(mw, NewRecorderOpt(rec))
+	hc := httpcli.NewFactory(mw, httpcli.NewCachedTransportOpt(rcache.NewWithTTL("http", 604800), true))
```
@davejrt it seems you've removed the recorder entirely here, we want the cached transport _as well as_ the `NewRecorderOpt(rec)`, I suspect many of the test failures have to do with this

bobheadxi replied:
The recorder is responsible for recording/using VCRs for tests

bobheadxi commented on internal/extsvc/github/testdata/vcr/TestListMembers/team_members.yaml:
```diff
@@ -1,75 +0,0 @@
-      - application/vnd.github.machine-man-preview+json
-      Content-Type:
-      - application/json; charset=utf-8
-    url: https://api.github.com/orgs/sourcegraph-vcr-repos/teams/private-access/members?page=1&per_page=100
```
@davejrt  I think these vcrs need to be added back, a group of the test failures are from `TestListMembers` which correspond to these VCRs. For example https://buildkite.com/sourcegraph/sourcegraph/builds/140457#2af03904-f6ee-40ab-ab1b-c77fd872c345/34-5992

```


\| t=2022-04-04T22:30:05+0000 lvl=dbug msg="TRACE github" host=api.github.com path=/orgs/sourcegraph-vcr-repos/teams/private-access/members code=404 duration=75.016086ms
--
  | \|     v3_test.go:596: request to http://github-proxy/orgs/sourcegraph-vcr-repos/teams/private-access/members?page=1&per_page=100 returned status 404: Not Found
```

I think this is a 404 on the recording being missing

bobheadxi commented on internal/repos/testdata/golden/TestRepositoryQuery_Do/doesnt-exceed-limit:
```diff
@@ -11,9 +11,9 @@
     "IsArchived": false,
     "IsLocked": false,
     "IsDisabled": false,
-    "ViewerPermission": "ADMIN",
-    "StargazerCount": 17729,
-    "ForkCount": 1101
```
@davejrt I think these VCR changes need to be reverted as well, they correspond almost exactly to the test failures https://buildkite.com/sourcegraph/sourcegraph/builds/140457#8434df98-2967-4531-97a2-710073a4cba5/183-515 

```


\|     github_test.go:766: (-want, +got):
--
  | \|           strings.Join({
  | \|           	... // 269 identical bytes
  | \|           	": false,\n    \"IsFork\": false,\n    \"IsArchived\": false,\n    \"IsLo",
  | \|           	"cked\": false,\n    \"IsDisabled\": false,\n    \"ViewerPermission\": \"",
  | \|         - 	"READ",
  | \|         + 	"ADMIN",
  | \|           	"\",\n    \"StargazerCount\": 1",
  | \|         - 	"929",
  | \|         + 	"772",
  | \|           	"9,\n    \"ForkCount\": 1",
  | \|         - 	"208",
  | \|         + 	"101",
  | \|           	"\n   },\n   \"Error\": \"\"\n  }\n ]",
  | \|           }, "")


```

davejrt commented:
So the remaining failure on this has me a little perplexed. I understand why in essence it's failing, because of the code I've removed which was explicitly creating a new cache...however. Given what was implemented for the other function here to use the httpcli backed by httpcache, I can't replicate the same intended behaviour. 

Given the difference in the way the tests are written I'm never seeing the functions being called from httpcache (using a local copy and directing my go.mod to this) and therefore the cache isn't being written to in the way I'd expect. 

Do I need to manipulate these tests to view this behaviour, and in actual fact it will work with "caching for free" or is there really something here that I've removed without properly considering the consequences. 

@jhchabran or @bobheadxi if you feel like tagging someone else here from repo-management then feel free but I'm obviously concerned with starting to spam random engineers.

davejrt commented on internal/extsvc/github/common_test.go:
```diff
@@ -74,65 +73,19 @@ func newTestClientWithAuthenticator(t *testing.T, auth auth.Authenticator, cli h
+	cli, save := newV3TestClient(t, "GetRepository")
+	defer save()
+
+	t.Run("first run", func(t *testing.T) {
```
Added in a second run

bobheadxi commented on internal/extsvc/github/testdata/vcr/GetRepository.yaml:
```diff
@@ -50,23 +50,102 @@ interactions:
+      - nosniff
+      X-Frame-Options:
+      - deny
+      X-From-Cache:
+      - "1"
```
judging from https://sourcegraph.com/github.com/gregjones/httpcache/-/blob/httpcache.go?L26-27&utm_source=raycast-sourcegraph this is good! Nice 😁

bobheadxi replied:
etag matches up as well

bobheadxi commented on internal/extsvc/github/testdata/vcr/GetRepository.yaml:
```diff
@@ -50,23 +50,102 @@ interactions:
       - nosniff
       X-Frame-Options:
       - deny
+      X-From-Cache:
+      - "1"
```
Odd that the first instance has this enabled as well - I'm guessing it's filling from a local redis instance? should be fine though

davejrt replied:
let me look, I did clear my local redis cache ( I think )

bobheadxi reviewed (approved):
Awesome work, love the negative diff 😁

davejrt commented:
closes https://github.com/sourcegraph/sourcegraph/pull/33219

indradhanush commented on internal/extsvc/github/v3_test.go:
```diff
@@ -462,49 +462,6 @@ func TestListOrganizations(t *testing.T) {
-		if err != nil {
-			t.Fatal(err)
-		}
-
-		if orgs == nil {
-			t.Fatal("expected orgs but got nil")
-		}
-
-		if len(orgs) != 100 {
-			t.Fatalf("expected 100 orgs but got %d", len(orgs))
-		}
-
-		if nextSince < 1 {
-			t.Fatalf("expected nextSince to be a positive int but got %v", nextSince)
-		}
-	})
-
-	t.Run("enterprise-integration-with-cache", func(t *testing.T) {
-		rcache.SetupForTest(t)
-		cli, save := newV3TestEnterpriseClient(t, "ListOrganizations")
-		defer save()
-
-		if cli.orgsCache == nil {
-			t.Fatal("expected orgsCache to be initialised but is nil")
-		}
-
-		hash := cli.auth.Hash()
-		expectedEtagKey := hash + "-orgs-etag-0"
-		expectedOrgsKey := hash + "-orgs-0"
-
-		// When starting from scratch, the cache should be empty.
-		if val, ok := cli.orgsCache.Get(expectedEtagKey); ok {
-			t.Fatalf("expected key %q to be empty in cache, but found %s", expectedEtagKey, val)
-		}
-
-		if val, ok := cli.orgsCache.Get(expectedOrgsKey); ok {
-			t.Fatalf("expected key %q to be empty in cache, but found %s", expectedOrgsKey, val)
-		}
-
-		// Make the API call. This should also populate the cache.
```
These test cases should not have been deleted as they test the caching behaviour based on the HTTP response headers. I created a draft PR here to indicate how this is failing at the moment here: https://github.com/sourcegraph/sourcegraph/pull/34004

indradhanush reviewed (commented):
Proposing to revert this here: https://github.com/sourcegraph/sourcegraph/pull/34002.

bobheadxi commented on internal/extsvc/github/v3.go:
```diff
@@ -526,40 +486,11 @@ func (c *V3Client) GetOrganization(ctx context.Context, login string) (org *OrgD
-		}
-
-		return updateOrgsCache(respState)
+	_, err = c.get(ctx, path, &orgs)
+	if err != nil {
+		return nil, -1, err
```
The httpcli cache is transparent, and it returns the cached result automatically based on etags - users of httpcli do not need to add any explicit handling

bobheadxi replied:
there is some discussion further up in the PR that discusses how the behaviour exhibits itself via VCRs

indradhanush replied:
I did some digging today and found what I thought should have been the fix in PR #34004 with this commit: 979aa65c36820a15ef2ae0c9e140425ed9dba53d. 

Then I thought of improving the test case for `TestListOrganizations` in the [next commit](127cce0ce92d61b4b7de5a529e744740aab6a9bb) by asserting the returned orgs list, which turns out to be empty. 

Previously the `ListOrganizations` API would _always_ return a list of orgs, either by fetching it from the API or by reading it from the cache. But that has now changed to return a list of orgs for the first request or nil for cached requests when the data has not changed upstream.

This made me curious to test out the `GetRepository` API as well and I think we have a regression here. I added the test case in the [next commit](20d579a915d848e5ca5a8ee3b7dd2cb63739848c) and noticed two debug log statements which seem to indicate the API call is made twice. 

I added the same test case in the PR #34002 the one to revert this one in [this](ebd3c4e9850314905f1f566015acb51f65a1d959) commit and the log statement appears only once after I setup `redis` for the test case. 

Here's a screenshot of this:

![image](https://user-images.githubusercontent.com/2682729/163975769-d804878b-d855-4f6a-9ee1-68b34da0e951.png)

ryanslade replied:
I *think* we expect it to make two calls but the second one will respond that the cache is good and won't count towards rate limiting. Can you try running after clearing the vcr and see what happens? It might also be worth checking the response code at the point where httpcli cache actually makes the request.

indradhanush replied:
@ryanslade I added some debug log statements for the status code and the request counter, and it seems we _are_ sending two requests in the current version on `main` vs one request with the older behavior:

![image](https://user-images.githubusercontent.com/2682729/163982389-1dff0013-2bae-4c52-b333-6a0549a4d0ad.png)

My next step is to pull up something like wireshark and figure out if we are actually making two network calls here or if this is still an edge case with how the caching layer works in `httpcache` and if we're missing something.

ryanslade replied:
I think that might be expected. The old code would get the repo from the redis cache on the second attempt and not bother making any network calls. The new code will always hit github, but by sending the cache headers if nothing has changed the second call is very cheap and github disregards it for rate limiting. 

This is just based on my understanding of what should happen, I haven't looked at the code.

indradhanush replied:
The second API request should return a `304 Not Modified` like in the case of the `/organizations` API. But I'm not sure `/repos/<owner>/<name>` supports conditional requests the same way the `/organizations` API does in which case it would be counting against the API rate limit?

ryanslade replied:
Correct, if the API doesn't support conditional requests then we'll be hitting it harder now.

indradhanush replied:
Hmm. I stand corrected. It seems like `/repos/<owner>/<name>` also returns an `etag` in which case the cache can take advantage of conditional requests for `GetRepository` as well. 

So I don't think this is a regression in the `GetRepository` API any more, although the expectations do change a bit, since if the repository is modified (a new commit) the API returns a 200OK and it counts against the rate limit vs in the older implementation the redis cache would expire only after 10 minutes for GitHub.com and after 30 seconds for GitHub enterprise instances. 

But I don't think this is a big deal and `GetRepository` is safe. :ok_hand:

indradhanush replied:
But I'm still surprised that the second API call returns a 200 OK instead of a 304 Not Modified. I tested this with curl and got a 304 for the second API request when I used the `etag` in the request header. Do you think this warrants additional investigation? :detective:

indradhanush replied:
Did some more debug logs to print rate limit information. It seems like we're not using the cached request the second time (look for values of `ratelimit-used` and `ratelimit-remaining` in the `DBUG` logs:

![image](https://user-images.githubusercontent.com/2682729/163988459-8baeb4d6-d38f-4fe2-a16e-09d4591662c9.png)

ryanslade replied:
It sounds like our caching may not be working as expected if curls returns a 304 but our code doesn't. I think it's worth investigating

jhchabran replied:
Just as passing thought, https://sourcegraph.com/github.com/gregjones/httpcache@901d90724c79/-/blob/httpcache.go?L190 is very confusing. That being said, we should not see the limit rate decreasing.

bobheadxi replied:
@indradhanush do you have Redis running when you run the tests locally?

bobheadxi replied:
https://github.com/sourcegraph/sourcegraph/pull/34004/files#r853189345

bobheadxi replied:
```
=== RUN   TestGetRepository
t=2022-04-19T15:02:43-0700 lvl=dbug msg="TRACE github" host=api.github.com path=/repos/sourcegraph/sourcegraph code=200 duration=483.520708ms
t=2022-04-19T15:02:43-0700 lvl=dbug msg=doRequest status="200 OK" x-ratelimit-remaining=4997
t=2022-04-19T15:02:43-0700 lvl=dbug msg="TRACE github" host=api.github.com path=/repos/sourcegraph/sourcegraph code=200 duration=3.931083ms
t=2022-04-19T15:02:43-0700 lvl=dbug msg=doRequest status="200 OK" x-ratelimit-remaining=4997
```
